File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # BFS level traversal
        if not root:
            return []
        ans = []
        queue = deque([root])
        while queue:
            sz = len(queue)
            rightNode = None
            for i in range(sz):
                rightNode = queue.popleft()
                if rightNode.left:
                    queue.append(rightNode.left)
                if rightNode.right:
                    queue.append(rightNode.right)
            if rightNode:
                ans.append(rightNode.val)
        return ans

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // find the first node where p and q are on the 2 sides
        // or the first node == p or == q
        return dfs(root, p ,q);
    }

    TreeNode* dfs(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == p || root == q) {
            return root;
        }
        int min = std::min(p->val, q->val);
        int max = std::max(p->val, q->val);
        if (root->val > min && root->val < max) {
            return root;
        } else if (root->val > max) {
            return dfs(root->left, p, q);
        } else {
            return dfs(root->right, p, q);
        }
    }

};
